Revise content to match new troubleshooting guidelines (#118033)
elastic/elasticsearch: docs/reference/troubleshooting/common-issues/task-queue-backlog.asciidoc

[[task-queue-backlog]]

> === Task queue backlog

Author: Agreed... except I did it this way initially and then thought "hmm, maybe for SEO I should cover both bases." Still, I agree and am glad for the second!

> === Task queue backlog

Reviewer: oo smart also not a bad plan - ok with either

*******************************
*Product:* Elasticsearch +

> *Deployment type:* Elastic Cloud (hosted or self-managed), self-managed +

Reviewer: given that this tutorial uses cat thread pool apis it's def not for serverless ... but probably for everything else

> *Deployment type:* Elastic Cloud (hosted or self-managed), self-managed +

Author: I wonder whether there's a case for just saying "All types except serverless" -- seems like that might be the case often enough?

> *Deployment type:* Elastic Cloud (hosted or self-managed), self-managed +

Reviewer: :2c: until we clean up our usage of elastic cloud it's better to be specific (to help later contributors)

> *Deployment type:* Elastic Cloud (hosted or self-managed), self-managed +

Reviewer: (or do me a favor and add it as a comment)

> *Deployment type:* Elastic Cloud (hosted or self-managed), self-managed +

Reviewer: yeah I think we need to see how some conversations around that official messaging goes - but agree we should be able to link to a source of truth. afraid of putting this list together now because it is actively being discussed

*Versions:* All
*******************************

A backlogged task queue can prevent tasks from completing and lead to an 
unhealthy cluster state. Contributing factors include resource constraints, 
a large number of tasks triggered at once, and long-running tasks.

[discrete]
[[diagnose-task-queue-backlog]]
==== Diagnose a backlogged task queue

> **Check the thread pool status**

Reviewer: I think ideally we'd have an introductory sentence before this and reformat these as child headings

A <<high-cpu-usage,depleted thread pool>> can result in
<<rejected-requests,rejected requests>>. 

> Use the <<cat-thread-pool,cat thread pool API> to monitor

Reviewer: borky link
```suggestion
Use the <<cat-thread-pool,cat thread pool API>> to monitor
```

active threads, queued tasks, rejections, and completed tasks:

[source,console]
----
GET /_cat/thread_pool?v&s=t,n&h=type,name,node_name,active,queue,rejected,completed
----

* Look for high `active` and `queue` metrics, which indicate potential bottlenecks.

> * Analyze whether thread pool issues are specific to a <<data-tiers,data tier>> or
> caused by uneven node resource utilization such as <<hotspotting,hot spotting>>.

Reviewer: uneven node resource utilization is synonymous with hot spotting. would rephrase this to not be "caused by uneven node resource utilization" but instead as something to look for. hack edit:

```suggestion
* Analyze whether thread pool issues are specific to a <<data-tiers,data tier>>.
* Analyze whether a specific node's thread pool is depleting faster than others. This can be caused by uneven node resource utilization, also known as <<hotspotting,hot spotting>>.
```

stray thoughts: 
- wonder if we should even introduce hot spotting here or leave it to the resolution
- wonder if "analyze" is too loaded of a word (overthinking here). people might think it's harder than it is, when it's really a "check if"
- little red flag in the back of my mind here - in some cases, we can already tie diagnostics to resolutions at this stage. should we be tying them together somehow? 
- is hotspotting missing from the recommendations?

[discrete]
[[diagnose-hot-thread]]
**Inspect hot threads on each node**

If a particular thread pool queue is backed up, periodically poll the
<<cluster-nodes-hot-threads,nodes hot threads API>> to gauge the thread's
progression and ensure it has sufficient resources:

[source,console]
----
GET /_nodes/hot_threads
----

**Identify long-running node tasks**

Long-running tasks can also cause a backlog. Use the <<tasks,task
management API>> to check for excessive `running_time_in_nanos` values:

[source,console]
----
GET /_tasks?pretty=true&human=true&detailed=true
----

> You can filter on a specific `action`, such as <<docs-bulk,bulk indexing>> or search-related tasks.

Reviewer: I think it's nice to explain why we suggest filtering on these (they tend to be long-running)

* Filter on <<docs-bulk,bulk index>> actions:
+
[source,console]
----
GET /_tasks?human&detailed&actions=indices:data/write/bulk
----

* Filter on search actions:
+
[source,console]
----
GET /_tasks?human&detailed&actions=indices:data/write/search
----

**Look for long-running cluster tasks**

Use the <<cluster-pending,cluster pending tasks API>> to identify delays
in cluster state synchronization: 

[source,console]
----
GET /_cluster/pending_tasks
----

Tasks with a high `timeInQueue` value are likely contributing to the backlog.

[discrete]
[[resolve-task-queue-backlog]]

> ==== Recommendations

Reviewer: Re: the recommendations quandary

I think that we shouldn't sacrifice clarity for following linguistic rules, so a noun is ok here.

**Increase available resources** 

<<reduce-cpu-usage>> or increase thread pool sizes. 

For example, the `force_merge` thread pool defaults to a single thread. 
Increasing the size to 2 in `elasticsearch.yml` might help reduce a backlog 
of force merge requests:

[source,yaml]
----

> thread_pool.force_merge.size: 2

Reviewer: unfortunately we need to figure out whether ESS supports this setting. suspect the answer is no

Reviewer: <img width="522" alt="image" src="https://github.com/user-attachments/assets/ff0b31ca-62d5-45cb-8e98-b106da5b3210">
correct

----

For more information, see <<settings>>.

**Cancel stuck tasks**

> If an active task's <<diagnose-hot-thread,hot thread>> shows no progress, consider canceling the task.

Reviewer: I wonder how we can connect a hot thread to a task. is that done with the cat thread pool api?

Reviewer: do we know how to cancel tasks

[discrete]
==== Resources

Related symptoms:

* <<high-cpu-usage,High CPU usage>>

> * <<rejected-requests,Rejected requests>>

Reviewer: add hot spotting?

GET /_nodes/hot_threads
----

Although the hot threads API response does not list the specific tasks running on a thread, 
it provides a summary of the thread's activities. You can correlate a hot threads response 
with a <<tasks,task management API response>> to identify any overlap with specific tasks. For 
example, if the hot threads response indicates the thread is `performing a search query`, you can 

> <<diagnose-task-queue-long-running-node-tasks,check for long-running search tasks>> via the task management API.

Reviewer: avoid latinisms

```suggestion
<<diagnose-task-queue-long-running-node-tasks,check for long-running search tasks>> using the task management API.
```

GET /_cluster/pending_tasks
----

Tasks with a high `timeInQueue` value are likely contributing to the backlog and might
need to be <<resolve-task-queue-backlog-stuck-tasks,canceled>>.

[discrete]
[[resolve-task-queue-backlog]]
==== Recommendations

After identifying problematic threads and tasks, resolve the issue by increasing resources or canceling tasks.

[discrete]
[[resolve-task-queue-backlog-resources]]
===== Increase available resources

> If tasks are progressing slowly, try reducing <<reduce-cpu-usage,CPU usage>>.

Reviewer: ```suggestion
If tasks are progressing slowly, try <<reduce-cpu-usage,reducing CPU usage>>.
```

In some cases, you might need to increase the thread pool size. For example, the `force_merge` thread pool defaults to a single thread. 
Increasing the size to 2 might help reduce a backlog of force merge requests.

[discrete]
[[resolve-task-queue-backlog-stuck-tasks]]
===== Cancel stuck tasks

If an active task's <<diagnose-task-queue-hot-thread,hot thread>> shows no progress, consider <<task-cancellation,canceling the task>>.

[discrete]
[[resolve-task-queue-backlog-hotspotting]]
===== Address hot spotting

> If a specific node's thread pool is problematic, try addressing 

Reviewer: `problematic` is a little loaded
```suggestion
If a specific node's thread pool is depleting faster than others, try addressing
```
